pkg/universe.dagger.io/CODING_STYLE.md

```diff
@@ -0,0 +1,203 @@
+
+```cue
+// good
+#WriteFile: {}
+
+// bad
+#writeFile: {}
+#write_file: {}
+```
+
+Choose `camelCase` for "public" fields, instead of `snake_case` or `"kebab-case"`.
+
+```cue
+actions: {
+    // good
+    goLint: ...
+
+    // bad
+    go_lint: ...
+    "go-lint": ...
+}
+```
+
+## Private names
+
+Choose `_#camelCase` for private definitions, that should only be used in current package.
+
+```cue
+// good
+_#mergeStructs: { ... }
+
+// bad, starts with lower case but not private
+#mergeStructs: { ... }
+
+// bad, starts with upper case but private
+_#MergeStructs: { ... }
+```
+
+Choose `_camelCase` for private fields (implementation details).
+
+```cue
+#Copy: {
+    // good, indicates internal implementation
+    _copy: core.#Copy
+
+    // bad, indicates it can be used directly
+    copy: core.#Copy
+}
+```
+
+## Definitions (`#`) for *schemas*, fields for concrete *implementations*
+
+```cue
+// good, defines a schema
+#Copy: {
+    // good, implements a schema
+    input: #Image
+    _copy: core.#Copy & { ... }
+}
+
+// bad, not a definition
+Copy: {
+    // bad, not using a visible or hidden field (according to intent)
+    #input: #Image
+    _#copy: core.#Copy & { ... }
+}
+
+// ok for mixin
+Mixin: { ... }
+copy: #Copy & Mixin & { ... }
+```
+
+## All public definitions and fields must be documented
+
+Other fields are optional. The main idea is that public fields should be documented for someone
+reading Dagger’s documentation in the browser. Everything else is for someone looking at the code.
+
+```yaml
+// Copy files into an image
+#Copy: {
+    // Image to receive copied files
+    input: #Image
+
+    // Files to copy
+    contents: dagger.#FS
+
+    _copy: ...
+
+    // Resulting image with added files
+    output: ...
+}
+```
+
+## Use verbs for action names, nouns otherwise
+
+Action definition names should be verbs (e.g., `#Build`, `#Run`, `#Push`),
+otherwise they should be nouns as much as possible (e.g., `#Image`, `#Secret`, `#Socket`).
+
+## No field abbreviations
+
+CUE fields are more explicit and intuitive without abbreviations (e.g, `target` or
+`destination` instead of `dest`).
+
+## Don’t inline scripts
+
+Avoid inlining scripts (e.g., *sh*, *py*, etc). Instead put them in their own files,
+with proper extension, and use `core.#Source` to import into CUE. This allows linting
+and avoids some limitations (script size, escaping).
+
+Some exceptions may apply:
+
+- The script is really short;
+- Interpolation is necessary (although there may be other solutions).
+
+```cue
+// good
+src: core.#Source & {
+    path: "./src"
+}
+run: bash.#Run & {
+    script: {
+        directory: src.output
+        filename: "setup.sh"
+    }
+}
+
+// ok
+run: bash.#Run & {
+    script: contents: "echo hello world"
+}
+
+// bad, better solution available
+let name = "world"
+run: bash.#Run & {
+    script: contents: """
+        ...
+        echo hello \(world)
+        ...
+        """
+}
+
+// good, use env var in script
```

Comment:
Maybe add a special category about this: People must avoid using interpolation in script and use environment instead.

Reply:
Good idea